data-prepper-plugins/common/src/main/java/com/amazon/dataprepper/plugins/buffer/BlockingBuffer.java

```diff
@@ -80,15 +92,29 @@ public BlockingBuffer(final String pipelineName) {
 
     @Override
     public void doWrite(T record, int timeoutInMillis) throws TimeoutException {
+        long nanos = TimeUnit.MILLISECONDS.toNanos(timeoutInMillis);
+        final int c;
         try {
-            boolean isSuccess = blockingQueue.offer(record, timeoutInMillis, TimeUnit.MILLISECONDS);
-            if (!isSuccess) {
-                throw new TimeoutException(format("Pipeline [%s] - Buffer is full, timed out waiting for a slot",
-                        pipelineName));
+            writeLock.lockInterruptibly();
+            final AtomicInteger count = this.numInflightRecords;
+            while (count.get() == this.bufferCapacity) {
+                if (nanos <= 0L) {
+                    throw new TimeoutException(format("Pipeline [%s] - Buffer is full, timed out waiting for a slot",
+                            pipelineName));
+                }
+                nanos = notFull.awaitNanos(nanos);
+            }
+            blockingQueue.offer(record);
+            c = count.incrementAndGet();
+            if (c < this.bufferCapacity) {
+                // Wake up other doWrite waiting threads due to previous full buffer state.
+                notFull.signal();
```

Comment:
should this be in the finally?

Reply:
This part largely refers to `LinkedListBlockingQueue.offer(E e, long timeout, TimeUnit unit)` in which there is only lock release in the finally block. TBH, I cannot tell the essential difference between placing it here and placing it in finally. Is there any good practice I missed here?

Reply:
So there is no functional difference here because it seems that there isn't a way for an error to be thrown. The practice is that if there is an action you want to make sure happens (such as unlocking a lock), then we want that to be done in the finally block. 

In addition, I'm a little unsure on why we need the write function to unblock threads. Is the checkpoint calling the notFull signal not sufficient here?

Reply:
> In addition, I'm a little unsure on why we need the write function to unblock threads. Is the checkpoint calling the notFull signal not sufficient here?

So currently the checkpoint API calls notFull.signal ONLY when the previous capacity is full. In theory, there could be more than one thread waiting for write into the buffer due to previous full capacity (although in practice looks like our source that executes write is always in the main thread). That is why I called notFull.signal() here to wake up another waiting thread if the capacity is not full yet. Otherwise those threads will never be waken up.

I guess alternatively, we could just invoke notFull.signal unconditionally in the checkpoint API. I am not sure which one is better in performance.

> Also, why not call the "signalNotFull()" function here if we need to call this signal?

so here we already hold the guarding lock while signalNotFull includes grabbing the lock.

Reply:
I think calling the signal in the checkpoint whether or not it was previously full makes more sense to me. The reason is that the write can only increase the buffer capacity. So it doesn't make sense to send the notFull signal there, because the buffer can never change from full to not full there.

Reply:
Thanks for the input here. I am trying to simplify the logic here by using semaphore as Jeff suggested. So this part might be rewritten.